From 2569f07ef72048466924b805ae10d98c81b8033a Mon Sep 17 00:00:00 2001
From: Zexin Fu <zexifu@iis.ee.ethz.ch>
Date: Tue, 5 May 2026 14:30:37 +0200
Subject: [PATCH] Followup fixes for RISC-V debug support: WFI nop in dbg mode,
 EBREAK enters dbg mode, DRET restores prv from dcsr_prv, dcsr_prv resets to
 M-mode

---
 C910_RTL_FACTORY/gen_rtl/cp0/rtl/ct_cp0_iui.v  |  2 +-
 C910_RTL_FACTORY/gen_rtl/cp0/rtl/ct_cp0_regs.v | 10 ++++++++--
 2 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/C910_RTL_FACTORY/gen_rtl/cp0/rtl/ct_cp0_iui.v b/C910_RTL_FACTORY/gen_rtl/cp0/rtl/ct_cp0_iui.v
index 66bfb3f..6b637f8 100644
--- a/C910_RTL_FACTORY/gen_rtl/cp0/rtl/ct_cp0_iui.v
+++ b/C910_RTL_FACTORY/gen_rtl/cp0/rtl/ct_cp0_iui.v
@@ -1440,7 +1440,7 @@ assign inst_csr_wr = cp0_select && iui_privilege && iui_inst_csr
                   && !iui_inst_ro;
 
 //low power insctuction
-assign inst_lpmd   = cp0_select && iui_privilege && iui_inst_wfi;
+assign inst_lpmd   = cp0_select && iui_privilege && iui_inst_wfi && !iui_d_mode;
 
 //instruction type singel for flush and iu special generation
 //ignore psr s bit only indicate insctuction type information
diff --git a/C910_RTL_FACTORY/gen_rtl/cp0/rtl/ct_cp0_regs.v b/C910_RTL_FACTORY/gen_rtl/cp0/rtl/ct_cp0_regs.v
index b3a0074..8d89e0f 100644
--- a/C910_RTL_FACTORY/gen_rtl/cp0/rtl/ct_cp0_regs.v
+++ b/C910_RTL_FACTORY/gen_rtl/cp0/rtl/ct_cp0_regs.v
@@ -2730,7 +2730,9 @@ always @( rtu_cp0_expt_vld_no_dbg
        or mpp[1:0]
        or iui_regs_inst_sret
        or mdeleg_vld
-       or iui_regs_inst_mret)
+       or iui_regs_inst_mret
+       or iui_regs_inst_dret
+       or dcsr_prv[1:0])
 begin
   if(rtu_cp0_expt_vld_no_dbg && !mdeleg_vld)
     pm_wdata[1:0] = 2'b11;
@@ -2740,6 +2742,8 @@ begin
     pm_wdata[1:0] = mpp[1:0];
   else if(iui_regs_inst_sret)
     pm_wdata[1:0] = {1'b0, sstatus_spp};
+  else if(iui_regs_inst_dret)
+    pm_wdata[1:0] = dcsr_prv[1:0]; // DRET instruction
   else
     pm_wdata[1:0] = pm[1:0];
 // &CombEnd; @1900
@@ -3886,6 +3890,8 @@ begin
     debug_mode_q   <= 1'b1;
   end else if(debug_req_i) begin
     debug_mode_q   <= 1'b1;
+  end else if(is_vld_ebreak_inst_i && !debug_mode_q) begin
+    debug_mode_q   <= 1'b1;
   end else begin
     debug_mode_q   <= debug_mode_q;
   end
@@ -3952,7 +3958,7 @@ begin
   if(!cpurst_b) begin
     dcsr_cause     <= 3'b0;
     dcsr_v         <= 1'b0;
-    dcsr_prv       <= 2'b0;
+    dcsr_prv       <= 2'b11; // M-Mode
   end else if(dcsr_local_en) begin
     dcsr_cause     <= iui_regs_src0[8:6];
     dcsr_v         <= iui_regs_src0[5];
-- 
2.43.7

